File: python/41.缺失的第一个正数.py
#
# @lc app=leetcode.cn id=41 lang=python3
#
# [41] 缺失的第一个正数
#
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        if not nums: return 1

        for i,num in enumerate(nums):
            # 如果在 [0,len(nums)-1]的范围内，
            # 并且 num不在 nums 的第 num 个位置
            # 就 交换位置
            while 0 <= num-1 < len(nums) and nums[num-1] != num:
                tmp = num-1
                num, nums[i], nums[tmp] = nums[tmp], nums[tmp], nums[i]
        for i in range(len(nums)):
            if nums[i] != i+1:
                return i+1
        return len(nums)+1

File: cpp/41.缺失的第一个正数.cpp
/*
 * @lc app=leetcode.cn id=41 lang=cpp
 *
 * [41] 缺失的第一个正数
 *
 * https://leetcode-cn.com/problems/first-missing-positive/description/
 *
 * algorithms
 * Hard (40.23%)
 * Likes:    668
 * Dislikes: 0
 * Total Accepted:    75.6K
 * Total Submissions: 188K
 * Testcase Example:  '[1,2,0]'
 *
 * 给你一个未排序的整数数组，请你找出其中没有出现的最小的正整数。
 * 
 * 
 * 
 * 示例 1:
 * 
 * 输入: [1,2,0]
 * 输出: 3
 * 
 * 
 * 示例 2:
 * 
 * 输入: [3,4,-1,1]
 * 输出: 2
 * 
 * 
 * 示例 3:
 * 
 * 输入: [7,8,9,11,12]
 * 输出: 1
 * 
 * 
 * 
 * 
 * 提示：
 * 
 * 你的算法的时间复杂度应为O(n)，并且只能使用常数级别的额外空间。
 * 
 */

#include "leetcode.h"
// @lc code=start
class Solution
{
public:
  int firstMissingPositive(vector<int>& nums)
  {
    for (int i = 0; i < nums.size(); i++) {
      if (i + 1 == nums[i]) continue;
      int x = nums[i];
      while (x >= 1 && x <= nums.size() && x != nums[x - 1]) {
        swap(x, nums[x - 1]);
      }
    }
    for (int i = 0; i < nums.size(); i++) {
      if (i + 1 != nums[i]) return i + 1;
    }
    return nums.size() + 1;
  }
};
// @lc code=end
